sqlmesh/core/reference.py

```diff
@@ -0,0 +1,113 @@
+from __future__ import annotations
+
+import typing as t
+from collections import defaultdict, deque
+
+from pydantic import validator
+from sqlglot import exp
+
+from sqlmesh.utils.errors import ConfigError, SQLMeshError
+from sqlmesh.utils.pydantic import PydanticModel
+
+if t.TYPE_CHECKING:
+    from sqlmesh.core.model import Model
+
+
+class Reference(PydanticModel, frozen=True):
+    expression: exp.Expression
+    dialect: str
+    unique: bool = False
+    _name: str = ""
+
+    @property
+    def columns(self) -> t.List[str]:
+        expression = self.expression
+        if isinstance(expression, exp.Alias):
+            expression = expression.this
+        expression = expression.unnest()
+        if isinstance(expression, (exp.Array, exp.Tuple)):
+            return [e.output_name for e in expression.expressions]
+        return [expression.output_name]
+
+    @validator("expression", pre=True)
+    def _expression_validator(cls, v: t.Any, values: t.Dict[str, t.Any]) -> exp.Expression:
+        from sqlmesh.core.model.common import parse_expression
+
+        expression = parse_expression(v, values)
+
+        if not isinstance(expression, exp.Expression):
+            raise ConfigError(
+                f"Reference '{v}' must be a single expression. For compound keys create a tuple eg: (a, b) as c"
+            )
+
+        if isinstance(expression, exp.Identifier):
+            expression = exp.column(expression)
+        return expression
+
+    @property
+    def name(self) -> str:
+        if not self._name:
+            keys = []
+
+            if isinstance(self.expression, (exp.Tuple, exp.Array)):
+                for e in self.expression.expressions:
+                    if not e.output_name:
+                        raise ConfigError(
+                            f"Reference '{e}' must have an inferrable name or explicit alias."
```

Comment:
Do we even allow arbitrary expressions in references? Should we validate this in the expression itself and make sure that only column identifiers are passed?

Reply:
yea, we allow compound ones like (a, b) as compound_ref